Preprocessor.idr

```diff
@@ -0,0 +1,123 @@
+module Preprocessor
+
+import System.File.ReadWrite
+import System.File.Handle
+import System
+import Control.Monad.Error.Either
+import Control.Monad.Trans
+import Data.List1
+import Data.List
+import Data.String
+import Data.String.Extra
+
+record MimeRecord where
+  constructor MkMimeRecord
+  mime_type : String
+  mime_exts : List String
+
+record_to_declaration_name : MimeRecord -> String
+record_to_declaration_name mime = concat $ map go $ unpack mime.mime_type
+  where
+    go : Char -> String
+    go '+' = "_PLUS_"
+    go '.' = "_DOT_"
+    go chr = if isAlphaNum chr then cast (toUpper chr) else "_"
+
+tail : String -> String
+tail "" = ""
+tail xs = assert_total (strTail xs)
+
+cap_first : String -> String
+cap_first str =
+  case fromList (unpack str) of
+    Just (h ::: t) => pack $ forget (toUpper h ::: t)
+    Nothing => ""
+
+record_to_sub_type : MimeRecord -> String
+record_to_sub_type mime = tail $ snd $ break ('/' ==) mime.mime_type
+
+record_to_main_type : MimeRecord -> String
+record_to_main_type mime = concat $ map cap_first $ split (not . isAlphaNum) $ fst $ break ('/' ==) mime.mime_type
+
+parse_line : String -> Maybe MimeRecord
+parse_line line = do
+  let line = remove_comment line
+  (type ::: exts) <- List1.fromList $ filter (/= "") $ forget $ split isSpace line
+  pure (MkMimeRecord type exts)
+  where
+    remove_comment : String -> String
+    remove_comment str = trim (if isPrefixOf "#" str then tail str else str)
+
+read_line_or_eof : HasIO io => File -> EitherT FileError io (Maybe String)
+read_line_or_eof file = do
+  False <- lift (fEOF file)
+  | True => pure Nothing
+  map Just $ MkEitherT $ fGetLine file
+
+read_from_file : HasIO io => String -> io (Either FileError (List MimeRecord))
+read_from_file file = runEitherT $ do
+  handle <- MkEitherT {m=io} (openFile file Read)
+  skip_comments handle
+  go [] handle
```

Comment:
Because of the recursion the list of mime types are in the opposite order as the input.
This order change makes hard to compare the end result and spot differences.

Can you reverse the result so the ordering is the same as in the input file?

Reply:
I've addressed this by adding `reverse` before `go`. Probably not the best in terms of performance but I think it works good enough.